byterun/win32.c

```diff
@@ -914,3 +918,58 @@ void caml_restore_win32_terminal(void)
   if (startup_codepage != 0)
     SetConsoleOutputCP(startup_codepage);
 }
+
+/* Detect if a named pipe corresponds to MSYS/Cygwin tty: see
+
+   https://github.com/mirror/newlib-cygwin/blob/master/winsup/cygwin/dtable.cc#L932
+*/
+typedef
+BOOL (WINAPI *tGetFileInformationByHandleEx)(
+  HANDLE                    hFile,
+  FILE_INFO_BY_HANDLE_CLASS FileInformationClass,
+  LPVOID                    lpFileInformation,
+  DWORD                     dwBufferSize
+);
+
+static int caml_win32_ismsystty(int fd)
+{
+  char buffer[1024];
+  FILE_NAME_INFO * nameinfo = (FILE_NAME_INFO *) buffer;
+  HMODULE hModKernel32;
+  tGetFileInformationByHandleEx pGetFileInformationByHandleEx;
+
+  /* check if fd is a pipe */
+  HANDLE h = (HANDLE) _get_osfhandle(fd);
+  if (GetFileType(h) != FILE_TYPE_PIPE)
+    return 0;
+
+  hModKernel32 = GetModuleHandle(L"KERNEL32.DLL");
+  pGetFileInformationByHandleEx =
+    (tGetFileInformationByHandleEx)GetProcAddress(hModKernel32, "GetFileInformationByHandleEx");
+
+  if (pGetFileInformationByHandleEx == NULL)
+    return 0;
+
+  /* get pipe name */
+  if (! pGetFileInformationByHandleEx(h, FileNameInfo, buffer, sizeof(buffer) - sizeof(WCHAR)))
+    return 0;
+
+  nameinfo->FileName[nameinfo->FileNameLength / sizeof(WCHAR)] = L'\0';
+
+  /* check if this could be a msys pty pipe ('msys-XXXX-ptyN-XX')
+     or a cygwin pty pipe ('cygwin-XXXX-ptyN-XX') */
+  if ((wcsstr(nameinfo->FileName, L"msys-") ||
+       wcsstr(nameinfo->FileName, L"cygwin-")) && wcsstr(nameinfo->FileName, L"-pty"))
+    return 1;
+
+  return 0;
+}
+
+CAMLexport int caml_win32_isatty(int fd)
+{
+  DWORD lpMode;
+  HANDLE hFile = (HANDLE)_get_osfhandle(fd);
+  return (hFile != INVALID_HANDLE_VALUE &&
+          GetFileType(hFile) == FILE_TYPE_CHAR &&
+          GetConsoleMode(hFile, &lpMode)) || caml_win32_ismsystty(fd);
+}
```

Comment:
I had to read this twice to make sure the `&&` and `||` were in the right places.  What about:
```
if (hFile == INVALD_HANDLE_VALUE) return 0;
if (GetFileType(hFile) == FILE_TYPE_CHAR && GetConsoleMode(hFile, &lpMode)) return 1;
return caml_win32_ismsystty(h);
```

Reply:
Yes, the suggested version is indeed clearer.